src/controllers/AddAssetsManager.js

```diff
@@ -0,0 +1,234 @@
+    const unApprovedAssetsCount = Object.keys(unApprovedAssets).length
+    this.store.updateState({ unApprovedAssets, unApprovedAssetsCount })
+  }
+
+  async validateWatchAssetParams(assetParams, web3) {
+    // log.debug('asset params', assetParams)
+    const supported_asset_types = [CONTRACT_TYPE_ERC20, CONTRACT_TYPE_ERC721, CONTRACT_TYPE_ERC1155]
+    if (!supported_asset_types.includes(assetParams.type.toLowerCase())) {
+      throw ethErrors.rpc.invalidParams('Invalid watch asset params: only erc20/erc721/erc1155 asset types are supported.')
+    }
+    const { address } = assetParams.options || {}
+    if (!address) throw ethErrors.rpc.invalidParams('Invalid watch asset params: asset address is required.')
+    const chainId = await web3.eth.getChainId()
+    const isValidAddress = await validateContractAddress(web3, address, chainId)
+    if (!isValidAddress) throw ethErrors.rpc.invalidParams(`Invalid watch asset params: Invalid asset address ${address}`)
+  }
+
+  async normalizeWatchAssetParams(assetId, assetParams, providerConfig, web3) {
+    const nft_contract_standard = [CONTRACT_TYPE_ERC721, CONTRACT_TYPE_ERC1155]
+    let finalParams = {}
+    if (assetParams.type.toLowerCase() === CONTRACT_TYPE_ERC20) {
+      const { address, decimals, symbol } = assetParams.options || {}
+      if (!address) throw ethErrors.rpc.invalidParams('Invalid watch asset params: asset address is required.')
+      const tokenHandler = new TokenHandler({ address: address.toLowerCase(), web3 })
+      const options = assetParams.options || {}
+      if (!decimals) options.decimals = await tokenHandler.getDecimals()
+      if (!symbol) options.symbol = await tokenHandler.getSymbol()
+      const name = await tokenHandler.getName()
+      const metadata = {
+        network: providerConfig.host,
+        name,
+      }
+      finalParams = {
+        id: assetId,
+        ...assetParams,
+        options: {
+          ...assetParams.options,
+          ...options,
+        },
+        metadata,
+      }
+      // return finalParams
+    }
+    if (nft_contract_standard.includes(assetParams.type.toLowerCase())) {
+      const { address, id, image, name, balance } = assetParams.options || {}
+      const userAddress = this.prefsController.store.getState().selectedAddress
+      log.info(userAddress)
+      if (!address) throw ethErrors.rpc.invalidParams('Invalid watch asset params: asset address is required.')
+      const nftHandler = new NftHandler({ userAddress, tokenId: id, address: address.toLowerCase(), web3 })
+      const nft_standard = await nftHandler.checkNftStandard().standard
+      const options = assetParams.options || {}
+      if (!image) options.image = await nftHandler.getCollectibleTokenURI(options.id, nft_standard)
+      if (!name) options.name = await nftHandler.getAssetName()
+      if (!balance) options.balance = await nftHandler.fetchNftBalance()
+      let metadata = await nftHandler.getNftMetadata(nft_standard)
```

Comment:
do we need this metadata? I don't think we are using this metadata anywhere.

Reply:
we r using metadata for nft balance, image, etc

Reply:
you are already adding the balance and image to options, so why assign it again to metadata